hoomd/hpmc/shape_move.py

```diff
@@ -65,215 +72,152 @@ class Constant(ShapeMove):
         mc.shape["A"] = dict(vertices=[(1, 1, 1), (-1, -1, 1),
                                        (1, -1, -1), (-1, 1, -1)])
         # example callback
-        class ExampleCallback(hoomd.hpmc.shape_move.Callback):
-            def __call__(self, params):
+        class ExampleCallback:
+            def __init__(self):
+                default_dict = dict(sweep_radius=0, ignore_statistics=True)
+            def __call__(self, type, param_list):
                 # do something with params and define verts
-                return hoomd.hpmc._hpmc.PolyhedronVertices(verts)
-        python_move = hoomd.hpmc.shape_move.Python(callback=ExampleCallback,
-                                                   params={'A': [1.0]},
-                                                   param_move_probability=1.0)
+                return dict("vertices":verts, **self.default_dict))
+        python_move = hpmc.shape_move.PythonShapeMove()
+        python_move.callback = ExampleCallback()
 
     Attributes:
+        callback (`callable`): The python function that will be called to perform
+            custom shape moves on arbitrary shape parameters. The function must
+            take the particle type and a list of parameters as arguments and
+            return a dictionary with the shape definition whose keys must match
+            the shape definition of the integrator:
 
-        callback (Callback): The python class that will be called
-            to update the particle shapes
+                ``fun(typeid, param_list) -> dict``
 
-        params (dict): Dictionary of types and the corresponding list
-            of initial parameters to pass to the callback
-            (ex: {'A' : [1.0], 'B': [0.0]})
+        params (`TypeParameter` [``particle type``, `list`]): List of tunable
+            parameters to be updated.
 
-        param_move_probability (float): Average fraction of parameters to change during
-            each shape move
+        move_probability (`float`): Average fraction of shape parameters to
+            change each timestep.
     """
 
-    def __init__(self, callback, params, param_move_probability):
-        # TODO: params should be implemented as TypeParameter
-        param_dict = ParameterDict(
-            callback=Callback,
-            params=dict(params),
-            param_move_probability=float(param_move_probability))
-        param_dict["callback"] = callback
-        self._param_dict.update(param_dict)
+    _suported_shapes = {
+        'ConvexPolyhedron', 'ConvexSpheropolyhedron', 'Ellipsoid'
+    }
 
-    def _attach(self):
-        integrator = self._simulation.operations.integrator
-        if not isinstance(integrator, integrate.HPMCIntegrator):
-            raise RuntimeError("The integrator must be a HPMC integrator.")
-        if not integrator._attached:
-            raise RuntimeError("Integrator is not attached yet.")
+    def __init__(self, move_probability=1):
```

Comment:
I kinda like the idea of keeping the same constructor across all shape moves, and setting specific attributes separately. Its not a strong preference and can go either way, though.

Reply:
I think of this like I do `Triggers` rather than `md.pair.Pair` objects. Typically attributes that are not per type or default per-type parameters are set on construction. Also, `Constant` has a different signature than the others.